collections/src/main/scala/strawman/collection/immutable/ChampHashMap.scala

```diff
@@ -0,0 +1,680 @@
+  *  @define coll immutable champ hash map
+  */
+
+@SerialVersionUID(2L)
+private[immutable] sealed case class ChampHashMap[K, +V](val rootNode: MapNode[K, V], val cachedJavaHashCode: Int, val cachedSize: Int)
+  extends Map[K, V]
+    with MapOps[K, V, ChampHashMap, ChampHashMap[K, V]]
+    with StrictOptimizedIterableOps[(K, V), Iterable /* ChampHashMap */, ChampHashMap[K, V]]
+    with Serializable {
+
+  override def iterableFactory = List
+
+  override def mapFactory = ChampHashMap
+
+  override protected[this] def mapFromIterable[K2, V2](it: collection.Iterable[(K2, V2)]): ChampHashMap[K2, V2] = ChampHashMap.from(it)
+
+  override protected[this] def fromSpecificIterable(coll: collection.Iterable[(K, V)]): ChampHashMap[K, V] = ChampHashMap.from(coll)
+
+  override protected[this] def newSpecificBuilder(): Builder[(K, V), ChampHashMap[K, V]] = ChampHashMap.newBuilder()
+
+  override def knownSize: Int = cachedSize
+
+  override def iterator(): Iterator[(K, V)] = new MapKeyValuePairIterator[K, V](rootNode)
+
+  override def get(key: K): Option[V] = rootNode.get(key, computeHash(key), 0)
+
+  override def updated[V1 >: V](key: K, value: V1): ChampHashMap[K, V1] = {
+    val effect = MapEffect[K, V1]()
+    val keyHash = computeHash(key)
+    val newRootNode = rootNode.updated(key, value, keyHash, 0, effect)
+
+    if (effect.isModified) {
+      if (effect.hasReplacedValue) {
+        ChampHashMap(newRootNode, cachedJavaHashCode, cachedSize)
+      } else {
+        ChampHashMap(newRootNode, cachedJavaHashCode + keyHash, cachedSize + 1)
+      }
+    }
+    else this
+  }
+
+  override def remove(key: K): ChampHashMap[K, V] = {
+    val effect = MapEffect[K, V]()
+    val keyHash = computeHash(key)
+    val newRootNode = rootNode.removed(key, keyHash, 0, effect)
+
+    if (effect.isModified)
+      ChampHashMap(newRootNode, cachedJavaHashCode - keyHash, cachedSize - 1)
+    else this
+  }
+
+  override def empty: ChampHashMap[K, V] = ChampHashMap.empty
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case map: ChampHashMap[K, V] =>
+        (this eq map) ||
+          (this.cachedSize == map.cachedSize) &&
+            (this.cachedJavaHashCode == map.cachedJavaHashCode) &&
+            (this.rootNode == map.rootNode)
+      case _ => super.equals(that)
+    }
+
+  override def hashCode(): Int = collection.Set.unorderedHash(toIterable, "Map".##)
+
+}
+
+object MapNode {
+
+  private val EmptyMapNode = new BitmapIndexedMapNode(0, 0, Array.empty)
+
+  def empty[K, V]: MapNode[K, V] =
+    EmptyMapNode.asInstanceOf[MapNode[K, V]]
+
+  val TupleLength = 2
+
+  val HashCodeLength = 32
+
+  val BitPartitionSize = 5
+
+  val BitPartitionMask = (1 << BitPartitionSize) - 1
+
+  val MaxDepth = ceil(HashCodeLength.asInstanceOf[Double] / BitPartitionSize).asInstanceOf[Int]
+
+  val SizeEmpty = 0
+
+  val SizeOne = 1
+
+  val SizeMoreThanOne = 2
+
+  def $mask(hash: Int, shift: Int): Int = (hash >>> shift) & BitPartitionMask
```

Comment:
This is not a good idea to start an identifier with the `$` character because the scala compiler also uses this prefix for compiler-generated identifiers.

Reply:
The `$` prefix was a workaround for compiler issues that I had, but maybe you can recommend me a better solution.

Without the `$` prefix, the statement `val mask = mask(elementHash, shift)` will fail compilation: 
```
[error] ChampHashSet.scala:160: recursive value mask needs type
[error]     val mask = mask(elementHash, shift)
[error]                ^
[error] one error found
```

Even when using a type annotation as in `val mask: Int = mask(elementHash, shift)`, the compiler cannot infer that I want to call the mask function:
```
[error] ChampHashSet.scala:160: Int does not take parameters
[error]     val mask: Int = mask(elementHash, shift)
[error]                         ^
[error] one error found
```

Any suggestions apart from fully qualifying the method (e.g., with `SetNode.mask`)?

Reply:
One solution is to use a different name for `$mask` but a name that has no dollar sign in it. Like `maskValue`, for instance. Otherwise the solution that consists in fully qualifying the method is good too.